### DISCUSSION TITLE
Auto layouting with node's DOM size Using d3-hierarchy

### DISCUSSION THREAD

**Discussion by vasukomuravelli** [xyflow/xyflow#2768] [Q&A]
### Please describe the feature that you want to propose

I'm using d3-hierarchy to auto-layout my nodes as shown in reactflow pro examples. But I need to get the node's DOM size instead of hardcoding the NODE_WIDTH and NODE_HEIGHT for layout. It's because the node size is dynamic depending on the content inside the node.

I'm trying to set ref to my custom node elements, grab the node DOM size after the initial rendering, and then use dagre to do auto-layout. However, because the rendering of the custom node is controlled inside react-flow, I can't find a way to set the ref.

Do you have any suggestions? Is there a way I can access the node DOM elements? @moklick. Thanks in advance!

**Comment by rpelorosso-sv** [comment]
@vasukomuravelli  Did you find any way to do this? I've read that ReactFlow renders the nodes on initialization and stores the width and height. And that when you use getNodes() you can get the width and height of the rendered nodes, but I've found that if I add a new node then that node reports wrong dimensions while the rest of the nodes are ok. If I then add a new one then the previous one reports the right size (and again the new one is wrong)

**Comment by mati-o** [comment]
@vasukomuravelli, @rpelorosso-sv I have stumbled upon the same problem, although I've used dagre.js to layout my nodes, I had the same issue of dynamic node size.

Firstly I guesstimated it according to the data size, but it was ugly and inaccurate, so being absolutely beginner in React, to my luck, I've found there's the `useLayoutEffect` [hook](https://react.dev/reference/react/useLayoutEffect) which allows one to get elements before the browser repaints, already having their actual size.

So, I've tried some hacks and was able to come with a solution:

1. First, we need to wrap our component in a `<ReactFlowProvider>` to be able to get the nodes from the ReactFlow Instance

```tsx
        <ReactFlowProvider>
            <MyFlowApp/>
        </ReactFlowProvider>
```

2. Then, in our component we use it

```tsx

const getLayoutedElements(nodes, edges) {
   ... your layouting logic
}

const updateSizes = (nodes, reactFlowInstance) => {   
    for (const node of nodes) {
        const reactNode = reactFlowInstance.getNode(node.id);
        if (reactNode && reactNode.width && reactNode.height) {
            const width = reactNode.width; // Add some fine-tuning to these if needed
            const height = reactNode.height;
            node.data.width = width;
            node.data.height = height;
        } else {
            // If there's still no size available, return false to retry on the next hook call
            return false;
        }
    }

    return true;
}

function MyFlowApp() {
    const [nodes, setNodes, onNodesChange] = useNodesState(layoutedNodes);
    const [doUpdateSize, setDoUpdateSize] = React.useState(true);
    
    // This is why we need to wrap our component in ReactFlowProvider
    const reactFlowInstance = useReactFlow()
    
    useLayoutEffect(() => {
        if (doUpdateSize) { // Do only if required, e.g. the first time or when data changes.
            if (updateSizes(nodes, reactFlowInstance)) {
                const {nodes: layoutedNodes, edges: layoutedEdges} = getLayoutedElements(nodes, edges);
                setNodes([...layoutedNodes]);
                setEdges([...layoutedEdges]);
  
                setDoUpdateSize(false);
            }
        }

    });
    
    // ... the rest of the component
}
```

Let me know if I can clarify anything and I will try :) although I have zero react experience

**Comment by DylanMerigaud** [comment]
The reliable way to get real node sizes for layout is to let React Flow measure them first, then feed those in instead of hardcoding NODE_WIDTH/HEIGHT. Render the nodes hidden, read the sizes off the `dimensions` change events (or `node.measured` once `useNodesInitialized()` is true), lay out with those, then reveal. That sidesteps the ref-measuring race entirely.

I ended up wrapping that into a hook plus a `layout(nodes, edges, { sizeOf })` where `sizeOf` returns the measured `{ width, height }` per node, in case it saves you the boilerplate: https://www.npmjs.com/package/react-flow-auto-layout. It's dagre rather than d3-hierarchy, but the measure-then-layout part is the same idea.
